### PR TITLE
[proto] Small optimization for gaussian_blur functional op

### PR DESCRIPTION
**Pull request by vfdev-5**
Description:
- use softmax and multiplication to speed-up gaussian blur op

```
[-------------- Gaussian Blur cpu torch.uint8 --------------]
                     |  gaussian_blur v1  |  gaussian_blur v2
1 threads: --------------------------------------------------
      (3, 400, 400)  |        2.78        |        2.65
6 threads: --------------------------------------------------
      (3, 400, 400)  |        1.08        |        1.01

Times are in milliseconds (ms).

[-------------- Gaussian Blur cuda torch.uint8 -------------]
                     |  gaussian_blur v1  |  gaussian_blur v2
1 threads: --------------------------------------------------
      (3, 400, 400)  |        342         |        227
6 threads: --------------------------------------------------
      (3, 400, 400)  |        259         |        227

Times are in microseconds (us).

[------------- Gaussian Blur cpu torch.float32 -------------]
                     |  gaussian_blur v1  |  gaussian_blur v2
1 threads: --------------------------------------------------
      (3, 400, 400)  |        2160        |        2080
6 threads: --------------------------------------------------
      (3, 400, 400)  |         917        |         840

Times are in microseconds (us).

[------------- Gaussian Blur cuda torch.float32 ------------]
                     |  gaussian_blur v1  |  gaussian_blur v2
1 threads: --------------------------------------------------
      (3, 400, 400)  |        232         |        200
6 threads: --------------------------------------------------
      (3, 400, 400)  |        231         |        200

Times are in microseconds (us).
```
[Code source](https://github.com/vfdev-5/tvapiv2_benchmarks/blob/master/check_gaussian_blur.py)

---

This PR may also help with flaky gaussian blur tests.

Context:

Gaussian blur eager vs jit tests are flaky for uint8 input. This may be due to the fact that we cast to float32, perform the opertation in float32 and cast back to uint8. We generate conv kernels (created using exp op) in float32 and maybe accumulating precision errors.

Refs: https://github.com/pytorch/vision/pull/6755

### CODE REVIEW

**Comment by vfdev-5** [comment]
Successful runs on ubuntu-latest:
- https://github.com/pytorch/vision/actions/runs/3241542307/attempts/1
- https://github.com/pytorch/vision/actions/runs/3241542307/attempts/2

File: test/test_prototype_transforms_consistency.py [hunk 1]
```diff
@@ -323,6 +323,9 @@ def __init__(
         ],
         # ElasticTransform needs larger images to avoid the needed internal padding being larger than the actual image
         make_images_kwargs=dict(DEFAULT_MAKE_IMAGES_KWARGS, sizes=[(163, 163), (72, 333), (313, 95)]),
+        # We updated gaussian blur kernel generation with a faster and numerically more stable version
+        # This brings float32 accumulation visible in elastic transform -> we need to relax consistency tolerance
+        closeness_kwargs={"rtol": 1e-1, "atol": 1},
```
Comment (pmeier) [inline test/test_prototype_transforms_consistency.py:328]: The `atol` part is problematic here since it basically renders the `float32` test useless: with `atol=1`, every float image within the interval `[0.0, 1.0]` will match regardless of a working kernel or not. I currently don't have a good answer for this problem. See #6750.

I see two options now:

1. Accept that we can't test `float32` for now.
2. Duplicate this whole configuration and set `make_images_kwargs=dict(..., dtypes=[torch.uint8])` and `make_images_kwargs=dict(..., dtypes=[torch.float32])` with the correct tolerances respectively.

File: torchvision/prototype/transforms/functional/_misc.py [hunk 2]
```diff
@@ -70,7 +87,18 @@ def gaussian_blur_image_tensor(
...
+    dtype = image.dtype if torch.is_floating_point(image) else torch.float32
+    kernel = _get_gaussian_kernel2d(kernel_size, sigma, dtype=dtype, device=image.device)
+    kernel = kernel.expand(image.shape[-3], 1, kernel.shape[0], kernel.shape[1])
+
+    image, need_cast, need_squeeze, out_dtype = _FT._cast_squeeze_in(image, [kernel.dtype])
+
+    # padding = (left, right, top, bottom)
+    padding = [kernel_size[0] // 2, kernel_size[0] // 2, kernel_size[1] // 2, kernel_size[1] // 2]
+    output = _FT.torch_pad(image, padding, mode="reflect")
+    output = _FT.conv2d(output, kernel, groups=output.shape[-3])
```
Comment (pmeier) [inline torchvision/prototype/transforms/functional/_misc.py:99]: These two are just imported from `torch`, right? If so, can we also do that here instead importing from the legacy transforms? 

Plus, regarding the `conv2d`, could we maybe implement a common `nd_conv2d` as proposed in https://github.com/pytorch/vision/pull/6726#issue-1402722696? Basically, `nd_conv2d` should handle the `_cast_squeeze_in` and `_cast_squeeze_out` so the actual kernel is cleaner. That way, all kernels that use `conv2d` now could use that utility and don't need to worry about shape or dtype.

File: torchvision/prototype/transforms/functional/_misc.py [hunk 3]
```diff
@@ -32,6 +33,22 @@ def normalize(
...
+    kernel1d = torch.softmax(-x.pow_(2), dim=0)
+    return kernel1d
+
+
+def _get_gaussian_kernel2d(
+    kernel_size: List[int], sigma: List[float], dtype: torch.dtype, device: torch.device
+) -> torch.Tensor:
+    kernel1d_x = _get_gaussian_kernel1d(kernel_size[0], sigma[0]).to(device, dtype=dtype)
+    kernel1d_y = _get_gaussian_kernel1d(kernel_size[1], sigma[1]).to(device, dtype=dtype)
+    kernel2d = kernel1d_y.unsqueeze(-1) * kernel1d_x
```
Comment (pmeier) [inline torchvision/prototype/transforms/functional/_misc.py:48]: Doesn't this have to be a matmul?

Comment (vfdev-5) [reply torchvision/prototype/transforms/functional/_misc.py:48]: This one is faster and gives required output

Comment (pmeier) [reply torchvision/prototype/transforms/functional/_misc.py:48]: Yeah, since we do something like `(N, 1) @ (1, N)` there are no additions and thus regular broadcasting works here. Thanks!

File: torchvision/prototype/transforms/functional/_misc.py (hunk 2 above)
Comment (vfdev-5) [reply torchvision/prototype/transforms/functional/_misc.py:99]: OK, I can import methods directly. 

As for `nd_conv(2d?)`, I do not think we always have a situation when we can hide _cast_squeeze_in, _cast_squeeze_out inside `nd_*` op... Anyway, let's see that elsewhere (not this PR).

File: test/test_prototype_transforms_consistency.py (hunk 1 above)
Comment (vfdev-5) [reply test/test_prototype_transforms_consistency.py]: How about this https://github.com/pytorch/vision/pull/6762/files#diff-bf36f728ca93b501ac1ae0dd786c891c9c389cab57e4843cbd5d3b93a1b95068R311 ?

**Review by pmeier — APPROVED** [review]
LGTM if CI is green!
